File: src/pathfinding/Pathfinder.ts
import { HexCell } from '../types';
import { HexGrid } from '../core/HexGrid';
import { HexCoordinates } from '../core/HexCoordinates';
import { UnitManager } from '../units/UnitManager';
import { PriorityQueue } from './PriorityQueue';
import { getMovementCost, isPassable } from './MovementCosts';

/**
 * Result of a pathfinding operation.
 */
export interface PathResult {
  /** The path from start to end (inclusive of both), empty if no path found */
  path: HexCell[];
  /** Total movement cost of the path */
  cost: number;
  /** Whether the destination is reachable */
  reachable: boolean;
}

/**
 * A* pathfinder for hex grids.
 * Finds optimal paths considering terrain costs, elevation, and unit obstacles.
 */
export class Pathfinder {
  constructor(
    private grid: HexGrid,
    private unitManager?: UnitManager
  ) {}

  /**
   * Find the optimal path between two cells using A* algorithm.
   *
   * @param start - Starting cell
   * @param end - Destination cell
   * @param options - Optional configuration
   * @returns PathResult with path, cost, and reachability
   */
  findPath(
    start: HexCell,
    end: HexCell,
    options: {
      /** Ignore units when pathfinding (useful for theoretical paths) */
      ignoreUnits?: boolean;
      /** Maximum cost to search (limits search area) */
      maxCost?: number;
    } = {}
  ): PathResult {
    const { ignoreUnits = false, maxCost = Infinity } = options;

    // Quick check: destination must be passable
    if (!isPassable(end)) {
      return { path: [], cost: Infinity, reachable: false };
    }

    // Quick check: destination can't have a unit (unless ignoring units)
    if (!ignoreUnits && this.unitManager) {
      const unitAtEnd = this.unitManager.getUnitAt(end.q, end.r);
      if (unitAtEnd) {
        return { path: [], cost: Infinity, reachable: false };
      }
    }

    // Same cell - trivial path
    if (start.q === end.q && start.r === end.r) {
      return { path: [start], cost: 0, reachable: true };
    }

    const frontier = new PriorityQueue<HexCell>();
    const cameFrom = new Map<string, HexCell | null>();
    const costSoFar = new Map<string, number>();

    const startKey = this.cellKey(start);
    const endKey = this.cellKey(end);

    frontier.enqueue(start, 0);
    cameFrom.set(startKey, null);
    costSoFar.set(startKey, 0);

    while (!frontier.isEmpty()) {
      const current = frontier.dequeue()!;
      const currentKey = this.cellKey(current);

      // Found destination
      if (currentKey === endKey) {
        return {
          path: this.reconstructPath(cameFrom, start, end),
          cost: costSoFar.get(endKey)!,
          reachable: true,
        };
      }

      // Explore neighbors
      for (const neighbor of this.grid.getNeighbors(current)) {
        // Skip if there's a unit (unless ignoring units)
        if (!ignoreUnits && this.unitManager) {
          const unitAtNeighbor = this.unitManager.getUnitAt(neighbor.q, neighbor.r);
          // Allow destination even if pathfinding toward a unit (for attack targeting)
          if (unitAtNeighbor && this.cellKey(neighbor) !== endKey) {
            continue;
          }
        }

        const moveCost = getMovementCost(current, neighbor);

        // Skip impassable terrain
        if (!isFinite(moveCost)) {
          continue;
        }

        const newCost = costSoFar.get(currentKey)! + moveCost;

        // Skip if exceeds max cost
        if (newCost > maxCost) {
          continue;
        }

        const neighborKey = this.cellKey(neighbor);

        if (!costSoFar.has(neighborKey) || newCost < costSoFar.get(neighborKey)!) {
          costSoFar.set(neighborKey, newCost);

          // A* priority = cost so far + heuristic estimate to goal
          const priority = newCost + this.heuristic(neighbor, end);
          frontier.enqueue(neighbor, priority);
          cameFrom.set(neighborKey, current);
        }
      }
    }

    // No path found
    return { path: [], cost: Infinity, reachable: false };
  }

  /**
   * Get all cells reachable from a starting cell within a movement budget.
   * Useful for showing movement range.
   *
   * @param start - Starting cell
   * @param movementPoints - Maximum movement cost allowed
   * @param options - Optional configuration
   * @returns Map of reachable cells to their movement cost
   */
  getReachableCells(
    start: HexCell,
    movementPoints: number,
    options: {
      ignoreUnits?: boolean;
    } = {}
  ): Map<HexCell, number> {
    const { ignoreUnits = false } = options;
    const reachable = new Map<HexCell, number>();
    const frontier = new PriorityQueue<HexCell>();
    const costSoFar = new Map<string, number>();

    const startKey = this.cellKey(start);
    frontier.enqueue(start, 0);
    costSoFar.set(startKey, 0);
    reachable.set(start, 0);

    while (!frontier.isEmpty()) {
      const current = frontier.dequeue()!;
      const currentKey = this.cellKey(current);
      const currentCost = costSoFar.get(currentKey)!;

      for (const neighbor of this.grid.getNeighbors(current)) {
        // Skip if there's a unit (unless ignoring)
        if (!ignoreUnits && this.unitManager) {
          const unitAtNeighbor = this.unitManager.getUnitAt(neighbor.q, neighbor.r);
          if (unitAtNeighbor) {
            continue;
          }
        }

        const moveCost = getMovementCost(current, neighbor);

        // Skip impassable
        if (!isFinite(moveCost)) {
          continue;
        }

        const newCost = currentCost + moveCost;

        // Skip if exceeds movement budget
        if (newCost > movementPoints) {
          continue;
        }

        const neighborKey = this.cellKey(neighbor);

        if (!costSoFar.has(neighborKey) || newCost < costSoFar.get(neighborKey)!) {
          costSoFar.set(neighborKey, newCost);
          frontier.enqueue(neighbor, newCost);
          reachable.set(neighbor, newCost);
        }
      }
    }

    return reachable;
  }

  /**
   * Check if a path exists between two cells.
   * More efficient than findPath if you only need yes/no.
   */
  hasPath(start: HexCell, end: HexCell, ignoreUnits = false): boolean {
    const result = this.findPath(start, end, { ignoreUnits });
    return result.reachable;
  }

  /**
   * Get the movement cost between two adjacent cells.
   * Returns Infinity if not adjacent or impassable.
   */
  getStepCost(from: HexCell, to: HexCell): number {
    // Check if adjacent
    const fromCoords = new HexCoordinates(from.q, from.r);
    const toCoords = new HexCoordinates(to.q, to.r);

    if (fromCoords.distanceTo(toCoords) !== 1) {
      return Infinity; // Not adjacent
    }

    return getMovementCost(from, to);
  }

  /**
   * Heuristic function for A* - hex distance.
   * Uses minimum possible cost (1) to ensure admissibility.
   */
  private heuristic(a: HexCell, b: HexCell): number {
    const coordsA = new HexCoordinates(a.q, a.r);
    const coordsB = new HexCoordinates(b.q, b.r);
    return coordsA.distanceTo(coordsB);
  }

  /**
   * Generate a unique key for a cell.
   */
  private cellKey(cell: HexCell): string {
    return `${cell.q},${cell.r}`;
  }

  /**
   * Reconstruct the path from the cameFrom map.
   */
  private reconstructPath(
    cameFrom: Map<string, HexCell | null>,
    start: HexCell,
    end: HexCell
  ): HexCell[] {
    const path: HexCell[] = [];
    let current: HexCell | null = end;

    while (current) {
      path.unshift(current);
      const key = this.cellKey(current);
      current = cameFrom.get(key) ?? null;
    }

    return path;
  }
}

File: results/csharp_converted/pathfinder.cs
using Godot;
using Godot.Collections;


//# A* pathfinder for hex grids.
//# Finds optimal paths considering terrain costs, elevation, and unit obstacles.

//# Matches web/src/pathfinding/Pathfinder.ts
[GlobalClass]
public partial class Pathfinder : Godot.RefCounted
{
	public Godot.HexGrid Grid;
	public Godot.UnitManager UnitManager;


	public override void _Init(Godot.HexGrid p_grid, Godot.UnitManager p_unit_manager = null)
	{
		Grid = p_grid;
		UnitManager = p_unit_manager;


		//# Find the optimal path between two cells using A* algorithm.

	}//# Returns dictionary with path, cost, and reachable status.
	public Dictionary FindPath(Godot.HexCell start, Godot.HexCell end, Dictionary options = new Dictionary{})
	{
		var ignore_units = options.Get("ignore_units", false);
		var max_cost = options.Get("max_cost", Mathf.Inf);
		var unit_type = options.Get("unit_type", null);


		// Quick check: destination must be passable for this unit type
		var dest_passable;
		if(unit_type != null)
		{
			dest_passable = MovementCosts.IsPassableForUnit(end, unit_type);
		}
		else
		{
			dest_passable = MovementCosts.IsPassable(end);
		}

		if(!dest_passable)
		{
			return new Dictionary{{"path", new Array{}},{"cost", Mathf.Inf},{"reachable", false},};
		}


		// Quick check: destination can't have a unit (unless ignoring units)
		if(!ignore_units && UnitManager != null)
		{
			var unit_at_end = UnitManager.GetUnitAt(end.Q, end.R);
			if(unit_at_end != null)
			{
				return new Dictionary{{"path", new Array{}},{"cost", Mathf.Inf},{"reachable", false},};
			}
		}


		// Same cell - trivial path
		if(start.Q == end.Q && start.R == end.R)
		{
			return new Dictionary{{"path", new Array{start, }},{"cost", 0},{"reachable", true},};
		}

		var frontier = PriorityQueue.New();
		var came_from = new Dictionary{};
		// cell_key -> HexCell
		var cost_so_far = new Dictionary{};

		// cell_key -> float
		var start_key = _CellKey(start);
		var end_key = _CellKey(end);

		frontier.Enqueue(start, 0);
		came_from[start_key] = null;
		cost_so_far[start_key] = 0;

		while(!frontier.IsEmpty())
		{
			var current = frontier.Dequeue();
			var current_key = _CellKey(current);


			// Found destination
			if(current_key == end_key)
			{
				return new Dictionary{
									{"path", _ReconstructPath(came_from, start, end)},
									{"cost", cost_so_far[end_key]},
									{"reachable", true},
									};
			}


			// Explore neighbors
			foreach(HexCell neighbor in Grid.GetNeighbors(current))
			{

				// Skip if there's a unit (unless ignoring units)
				if(!ignore_units && UnitManager != null)
				{
					var unit_at_neighbor = UnitManager.GetUnitAt(neighbor.Q, neighbor.R);

					// Allow destination even if pathfinding toward a unit (for attack targeting)
					if(unit_at_neighbor != null && _CellKey(neighbor) != end_key)
					{
						continue;
					}
				}


				// Calculate movement cost (domain-aware if unit type provided)
				var move_cost;
				if(unit_type != null)
				{
					move_cost = MovementCosts.GetMovementCostForUnit(current, neighbor, unit_type);
				}
				else
				{
					move_cost = MovementCosts.GetMovementCost(current, neighbor);
				}


				// Skip impassable terrain
				if(!Mathf.IsFinite(move_cost))
				{
					continue;
				}

				var new_cost = cost_so_far[current_key] + move_cost;


				// Skip if exceeds max cost
				if(new_cost > max_cost)
				{
					continue;
				}

				var neighbor_key = _CellKey(neighbor);

				if(!cost_so_far.ContainsKey(neighbor_key) || new_cost < cost_so_far[neighbor_key])
				{
					cost_so_far[neighbor_key] = new_cost;


					// A* priority = cost so far + heuristic estimate to goal
					var priority = new_cost + _Heuristic(neighbor, end);
					frontier.Enqueue(neighbor, priority);
					came_from[neighbor_key] = current;
				}
			}
		}


		// No path found
		return new Dictionary{{"path", new Array{}},{"cost", Mathf.Inf},{"reachable", false},};


		//# Get all cells reachable from a starting cell within a movement budget.

	}//# Returns dictionary of cell -> movement cost.
	public Dictionary GetReachableCells(Godot.HexCell start, double movement_points, Dictionary options = new Dictionary{})
	{
		var ignore_units = options.Get("ignore_units", false);
		var unit_type = options.Get("unit_type", null);

		var reachable = new Dictionary{};
		// HexCell -> float
		var frontier = PriorityQueue.New();
		var cost_so_far = new Dictionary{};

		// cell_key -> float
		var start_key = _CellKey(start);
		frontier.Enqueue(start, 0);
		cost_so_far[start_key] = 0;
		reachable[start] = 0;

		while(!frontier.IsEmpty())
		{
			var current = frontier.Dequeue();
			if(current == null)
			{
				break;
			}
			var current_key = _CellKey(current);
			var current_cost = cost_so_far.Get(current_key, 0.0);

			foreach(HexCell neighbor in Grid.GetNeighbors(current))
			{

				// Skip if there's a unit (unless ignoring)
				if(!ignore_units && UnitManager != null)
				{
					var unit_at_neighbor = UnitManager.GetUnitAt(neighbor.Q, neighbor.R);
					if(unit_at_neighbor != null)
					{
						continue;
					}
				}


				// Calculate movement cost (domain-aware if unit type provided)
				var move_cost;
				if(unit_type != null)
				{
					move_cost = MovementCosts.GetMovementCostForUnit(current, neighbor, unit_type);
				}
				else
				{
					move_cost = MovementCosts.GetMovementCost(current, neighbor);
				}


				// Skip impassable
				if(!Mathf.IsFinite(move_cost))
				{
					continue;
				}

				var new_cost = current_cost + move_cost;


				// Skip if exceeds movement budget
				if(new_cost > movement_points)
				{
					continue;
				}

				var neighbor_key = _CellKey(neighbor);

				if(!cost_so_far.ContainsKey(neighbor_key) || new_cost < cost_so_far[neighbor_key])
				{
					cost_so_far[neighbor_key] = new_cost;
					frontier.Enqueue(neighbor, new_cost);
					reachable[neighbor] = new_cost;
				}
			}
		}

		return reachable;
	}


	//# Check if a path exists between two cells.
	public bool HasPath(Godot.HexCell start, Godot.HexCell end, bool ignore_units = false)
	{
		var result = FindPath(start, end, new Dictionary{{"ignore_units", ignore_units},});
		return result["reachable"];
	}


	//# Get the movement cost between two adjacent cells.
	public double GetStepCost(Godot.HexCell from, Godot.HexCell to)
	{

		// Check if adjacent
		var from_coords = HexCoordinates.New(from.Q, from.R);
		var to_coords = HexCoordinates.New(to.Q, to.R);

		if(from_coords.DistanceTo(to_coords) != 1)
		{
			return Mathf.Inf;
		}

		// Not adjacent
		return MovementCosts.GetMovementCost(from, to);
	}


	//# Heuristic function for A* - hex distance.
	protected double _Heuristic(Godot.HexCell a, Godot.HexCell b)
	{
		var coords_a = HexCoordinates.New(a.Q, a.R);
		var coords_b = HexCoordinates.New(b.Q, b.R);
		return Float(coords_a.DistanceTo(coords_b));
	}


	//# Generate a unique key for a cell.
	protected String _CellKey(Godot.HexCell cell)
	{
		return "%d,%d" % new Array{cell.Q, cell.R, };
	}


	//# Reconstruct the path from the came_from map.
	protected Array<HexCell> _ReconstructPath(Dictionary came_from, Godot.HexCell start, Godot.HexCell end)
	{
		var path = new Array{};
		var current = end;

		while(current != null)
		{
			path.PushFront(current);
			var key = _CellKey(current);
			current = came_from.Get(key);
		}

		return path;
	}


}